### PR TITLE
Feature/issues11 Add set mixer levels API

### PR DESCRIPTION
**Pull request by krishna-1809**

### CODE REVIEW

File: include/dsAudio.h [hunk 1]
```diff
@@ -2063,6 +2063,30 @@ dsError_t  dsGetSecondaryLanguage(intptr_t handle, char* sLang);
...
+* @retval dsERR_OPERATION_NOT_SUPPORTED    - The attempted operation is not supported; e.g: source devices
+* @retval dsERR_OPERATION_FAILED           - The attempted operation has failed
+*
+* @pre dsAudioPortInit() must be called before calling this API
+*
+* @warning  This API is Not thread safe
+*
+*
+*/
+dsError_t dsSetAudioMixerLevels (intptr_t handle, dsAudioInput_t aInput, int volume);
```
Comment (hari22yuva) [inline include/dsAudio.h:2088]: Do we get the handle without calling dsGetAudioPort()? if not add dsGetAudioPort() to Precondition

File: include/dsAVDTypes.h [hunk 2]
```diff
@@ -779,6 +779,13 @@ typedef enum _dsDisplayMatrixCoefficients_t
     dsDISPLAY_MATRIXCOEFFICIENT_MAX            ///< Out of range
 } dsDisplayMatrixCoefficients_t;
 
+typedef enum _dsAudioInput_t
```
Comment (hari22yuva) [inline include/dsAVDTypes.h:782]: add @brief and @note Used in @link

File: include/dsAudio.h (hunk 1 above)
Comment (hari22yuva) [inline include/dsAudio.h:2088]: Do we have get function for mixer volume level?

**Review by hari22yuva — CHANGES_REQUESTED** [review]
Please address the review comments

File: include/dsAVDTypes.h (hunk 2 above)
Comment (krishna-1809) [reply include/dsAVDTypes.h:786]: Added

File: include/dsAudio.h (hunk 1 above)
Comment (krishna-1809) [reply include/dsAudio.h:2088]: There is no getAudioMixerLevels() as of now

File: include/dsAVDTypes.h (hunk 2 above)
Comment (srinivasgtl) [reply include/dsAVDTypes.h:786]: Please specify "MS12 only" as this mixer is used for MS12 only.

File: include/dsAudio.h [hunk 3]
```diff
@@ -2063,6 +2063,30 @@ dsError_t  dsGetSecondaryLanguage(intptr_t handle, char* sLang);
  */
 dsError_t dsGetHDMIARCPortId(int *portId);
 
+/**
+* @brief Sets the Mixer Volume level for the given input
```
Comment (srinivasgtl) [inline include/dsAudio.h:2067]: Describe clearly that this functionality is for HDMI based input alone and also provide bit more information on input stream (MS12).

Comment (krishna-1809) [reply include/dsAudio.h:2067]: This is not used HDMI Input . It is basically used for controlling volume levels or mixer levels for each of the Primary and System Input

File: include/dsAVDTypes.h (hunk 2 above)
Comment (krishna-1809) [reply include/dsAVDTypes.h:786]: Updated

**Review by srinivasgtl — APPROVED** [review]
ship it.

File: include/dsAudio.h (hunk 3 above)
Comment (hari22yuva) [reply include/dsAudio.h]: Can you define what do mean by Primary and System Input

File: include/dsAudio.h [hunk 4]
```diff
@@ -2063,6 +2063,30 @@ dsError_t  dsGetSecondaryLanguage(intptr_t handle, char* sLang);
...
+* @retval dsERR_OPERATION_NOT_SUPPORTED    - The attempted operation is not supported; e.g: source devices
+* @retval dsERR_OPERATION_FAILED           - The attempted operation has failed
+*
+* @pre dsAudioPortInit() and and dsGetAudioPort() must be called before calling this API
+*
+* @warning  This API is Not thread safe
+*
+*
+*/
+dsError_t dsSetAudioMixerLevels (intptr_t handle, dsAudioInput_t aInput, int volume);
```
Comment (hari22yuva) [inline include/dsAudio.h:2088]: is this API need to support both sink &  source ?

File: include/dsAudio.h (hunk 3 above)
Comment (krishna-1809) [reply include/dsAudio.h]: updated

File: include/dsAudio.h (hunk 4 above)
Comment (krishna-1809) [reply include/dsAudio.h:2088]: For now it is implemented for sink device.

Comment (hari22yuva) [reply include/dsAudio.h:2088]: it's not about the current implementation, if any new vendor need to support this API for sink & source

File: include/dsAVDTypes.h (hunk 2 above)
Comment (Ulrond) [reply include/dsAVDTypes.h:786]: remove -> _dsAudioInput_t it's not required

File: include/dsAudio.h [hunk 5]
```diff
@@ -2063,6 +2063,30 @@ dsError_t  dsGetSecondaryLanguage(intptr_t handle, char* sLang);
...
```
Comment (Ulrond) [inline include/dsAudio.h:2088]: why is volume an int and not a char since it's range is 0-100?

Comment (Ulrond) [inline include/dsAudio.h:2088]: function should mention ms12 in it's API name, because it's specific for MS12.

File: include/dsAudio.h [hunk 6]
```diff
@@ -2063,6 +2063,30 @@ dsError_t  dsGetSecondaryLanguage(intptr_t handle, char* sLang);
...
+* @param[in] handle  - Handle for the output Audio port
+* @param[in] aInput  - dsAudioInputPrimary / dsAudioInputSystem. Please refer ::dsAudioInput_t
+* @param[in] volume  - volume to be set (0 to 100)
+*
+* @return dsError_t                        - Status
+* @retval dsERR_NONE                       - Success
+* @retval dsERR_NOT_INITIALIZED            - Module is not initialised
+* @retval dsERR_INVALID_PARAM              - Parameter passed to this function is invalid
+* @retval dsERR_OPERATION_NOT_SUPPORTED    - The attempted operation is not supported; e.g: source devices
+* @retval dsERR_OPERATION_FAILED           - The attempted operation has failed
```
Comment (Ulrond) [inline include/dsAudio.h:2080]: Why would the operation fail? Specifically, what reason are you suggesting here why it wouldnt' work?

File: include/dsAudio.h [hunk 7]
```diff
@@ -2063,6 +2063,30 @@ dsError_t  dsGetSecondaryLanguage(intptr_t handle, char* sLang);
  */
 dsError_t dsGetHDMIARCPortId(int *portId);
 
+/**
+* @brief Sets the Mixer Volume level for the MS12 input
```
Comment (Ulrond) [inline include/dsAudio.h:2067]: Sets the Mixer Volume level for the MS12 input -> Why is this brief mentioning ms12, but the function defination does not?

Is the setAudoMixerLevels also for non ms12 streams?

**Review by Ulrond — CHANGES_REQUESTED** [review]
some questions

File: include/dsAudio.h (hunk 4 above)
Comment (Ulrond) [reply include/dsAudio.h:2088]: Where is the dsGetMs12AudioMixerLevels()?

Comment (Ulrond) [reply include/dsAudio.h:2088]: If this function is specifically for sink then the documenation should method this

File: include/dsAudio.h (hunk 7 above)
Comment (Ulrond) [reply include/dsAudio.h:2067]: dsSetMS12AudoMixerLevels

Comment (ddevad) [reply include/dsAudio.h:2067]: No this should be a generic mixer API.  In amlogic platform, it's going to MS12

File: include/dsAudio.h (hunk 5 above)
Comment (ddevad) [reply include/dsAudio.h:2088]: It's not MS12 specific

Comment (ddevad) [reply include/dsAudio.h:2088]: To make it inline with all other volume level APIs in DS Audio subcomponent.  And believe int makes sense?

File: include/dsAudio.h (hunk 4 above)
Comment (ddevad) [reply include/dsAudio.h:2088]: Get API is not required as AS is managing and persisting these settings.   At the moment, set API should be good enough for unblocking the feature

File: include/dsAudio.h (hunk 6 above)
Comment (krishna-1809) [reply include/dsAudio.h]: Made it dsERR_GENERAL

File: include/dsAudio.h (hunk 7 above)
Comment (krishna-1809) [reply include/dsAudio.h]: Updated as generic mixer and removed MS12 from header and enum briefs

File: include/dsAudio.h (hunk 4 above)
Comment (krishna-1809) [reply include/dsAudio.h:2088]: This API is specific to sink

Comment (hari22yuva) [reply include/dsAudio.h:2088]: please mention same in @brief

Comment (krishna-1809) [reply include/dsAudio.h:2089]: Addressed

**Review by hari22yuva — APPROVED** [review]
ship it

**Review by Ulrond — APPROVED** [review]
looks good to me

**Review by hari22yuva — APPROVED** [review]
Ship it

**Pull request merged**